### PR TITLE
Refactor waiting loop to use resourceVersion

### LINKED ISSUE #1287: Refactor watch functions to pass resourceVersion
Let me add a few thoughts after digging into watch API a bit.
> 2. The CPU is pegged
> 3. `time.After` creates a new timer on each invocation. The timers aren't garbage collected until they expire, resulting in a crippling memory leak.

That's a neat fix. Thanks for pointing that out. The go docs are pretty clear there.

>     1. The code retries even though the results channel is never recreated, meaning the wait will never succeed

Indeed you're correct here as well. With the fix for watcher [race condition](https://github.com/knative/client/commit/83b8635d85c528f4da86c95ef5c3d74aa8543a30#diff-2a2da1f17418a3a96fbb5303e3406e5ae60f72536097b23d56f7ab583456ac8a) we actually lost the ability to recreate the watcher.
Even though the wait loop checks for `!ok` channel and it's the only [place](https://github.com/knative/client/blob/main/pkg/wait/wait_for_ready.go#L165-L167) producing retry=true, it's ineffective without remove initialization code.

https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes
>When retrieving a collection of resources (either namespace or cluster scoped), the response from the server will contain a resourceVersion value that can be used to initiate a watch against the server. The server will return all changes (creates, deletes, and updates) that occur after the supplied resourceVersion. This allows a client to fetch the current state and then watch for changes without missing any updates. 

As pointed out by @markusthoemmes, the `resourceVersion` can be used to avoid races like that, see the k8s docs above.

Therefore I think the right fix for the race condition is to leverage resourceVersion in our watch request. I've done a bit of testing and it seems to behave as expected, even with synthetic timeout postponing the watcher creation the events are lost. That'd make room to bring back the correct retry behaviour when channel is closed. As an good exampple and inspiration client-go's [retrywatcher](https://github.com/kubernetes/client-go/blob/master/tools/watch/retrywatcher.go) is done the same way.

_Originally posted by @dsimansk in https://github.com/knative/client/issues/1263#issuecomment-802225259_

### PR DESCRIPTION
**Pull request by dsimansk**
## Changes

* Use `resourceVersion` in `*Delete` functions to get replay of events and prevent 
* Re-enable `retry` logic when watch channel is closed before timeout

## Reference

Fixes #1287 

/cc @rhuss

### CODE REVIEW

File: pkg/wait/wait_for_ready.go [hunk 1]
```diff
@@ -140,10 +143,14 @@ func (w *waitForReadyConfig) Wait(ctx context.Context, watcher watch.Interface,
...
-	ctx context.Context, watcher watch.Interface, start time.Time, timeoutTimer *time.Timer, errorWindow time.Duration, msgCallback MessageCallback,
-) (retry bool, timeoutReached bool, err error) {
+func (w *waitForReadyConfig) waitForReadyCondition(ctx context.Context, name string, initialVersion string, start time.Time,
+	timeoutTimer *time.Timer, errorWindow time.Duration, options Options, msgCallback MessageCallback) (retry bool, timeoutReached bool, err error) {
 
+	watcher, err := w.watchMaker(ctx, name, initialVersion, options.timeoutWithDefault())
+	if err != nil {
+		return false, false, err
+	}
+	defer watcher.Stop()
```
Comment (maximilien) [inline pkg/wait/wait_for_ready.go:153]: I can comment lines 150-152 and all tests passing. Similar on the other changes. My guess is that your tests are not covering the cases for failed timeout or similar?

File: pkg/wait/poll_watcher.go [hunk 2]
```diff
@@ -83,6 +83,18 @@ func NewWatcher(ctx context.Context, watchFunc watchF, c rest.Interface, ns stri
...
+func NewWatcherWithVersion(ctx context.Context, watchFunc watchF, c rest.Interface, ns string, resource string, name string, initialResourceVersion string, timeout time.Duration) (watch.Interface, error) {
+	native, err := nativeWatchWithVersion(ctx, watchFunc, name, initialResourceVersion, timeout)
+	if err == nil {
+		return native, nil
+	}
+	polling := &pollingWatcher{
+		c, ns, resource, name, timeout, make(chan bool), make(chan watch.Event), &sync.WaitGroup{},
+		newTickerPollInterval(time.Second), nativePoll(ctx, c, ns, resource, name)}
+	polling.start()
+	return polling, nil
```
Comment (maximilien) [inline pkg/wait/poll_watcher.go:95]: I can comment this line and all tests pass. So missing coverage.

File: pkg/eventing/v1/client.go [hunk 3]
```diff
@@ -241,25 +242,27 @@ func (c *knEventingClient) GetBroker(ctx context.Context, name string) (*eventin
...
+	return wait.NewWatcherWithVersion(ctx, c.client.Brokers(c.namespace).Watch, c.client.RESTClient(), c.namespace, "brokers", name, initialVersion, timeout)
 }
 
 // DeleteBroker is used to delete an instance of broker and wait for completion until given timeout
 // For `timeout == 0` delete is performed async without any wait
 func (c *knEventingClient) DeleteBroker(ctx context.Context, name string, timeout time.Duration) error {
+	broker, err := c.GetBroker(ctx, name)
+	if err != nil {
+		return err
+	}
```
Comment (maximilien) [inline pkg/eventing/v1/client.go:255]: I can comment lines 253-255 and all tests passing

File: pkg/serving/v1/client.go [hunk 4]
```diff
@@ -314,24 +314,28 @@ func (cl *knServingClient) ApplyService(ctx context.Context, modifiedService *se
 // Param `timeout` represents a duration to wait for a delete op to finish.
 // For `timeout == 0` delete is performed async without any wait.
 func (cl *knServingClient) DeleteService(ctx context.Context, serviceName string, timeout time.Duration) error {
+	service, err := cl.GetService(ctx, serviceName)
+	if err != nil {
+		return err
+	}
```
Comment (maximilien) [inline pkg/serving/v1/client.go:320]: Same as previous...

**Review by maximilien — CHANGES_REQUESTED** [review]
/ok-to-test
Left some feedback, especially related to some missing coverage. Of course we cannot cover all but at least see if you can solve one --- some are same in different locations.

**Comment by dsimansk** [comment]
> Left some feedback, especially related to some missing coverage. Of course we cannot cover all but at least see if you can solve one --- some are same in different locations.

Thanks, sure I'll take a pass on the test. The waiting loop is pretty tricky anyway, so more coverage is definitely good goal.

**Comment by rhuss** [comment]
@dsimansk Thanks a lot ! I won't be able to make review this week, but will have a look asap early next week.

**Comment by maximilien** [comment]
Sounds good. Thanks @dsimansk

**Review by navidshaikh — APPROVED** [review]
/lgtm

**Pull request merged**